### PR TITLE
Opencensus propagation stackdriver

### PR DESCRIPTION
**Pull request by ofrobots**
Introduces a package for Stackdriver propagation. Right now only v1 format is implemented.

TODO:
- [x] Add tests for generate.

### CODE REVIEW

File: packages/opencensus-propagation-stackdriver/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  /** Span ID */
+  spanId: string;
+  /** Options */
+  options?: number;
+}
+
+/**
+ * An transport and environment neutral API for getting request headers.
+ */
+export interface Getter {
```
Comment (kjin) [inline packages/opencensus-propagation-stackdriver/src/index.ts]: Maybe we should prefer names that have to do more with headers or propagation. `Getter` seems too generic of an interface name.

File: packages/opencensus-propagation-stackdriver/src/v1.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
Comment (kjin) [inline packages/opencensus-propagation-stackdriver/src/v1.ts]: Does it make sense to update the copyright?

File: packages/opencensus-propagation-stackdriver/test/test-v1.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+describe('extract', () => {
+  describe('valid inputs', () => {
+    it('should return expected values: cafef00d/667;o=1', () => {
+      const getter = {
+        getHeader() {
+          return 'cafef00d/667;o=1';
+        }
+      };
+      const result = notNull(extract(getter));
+      assert.strictEqual(result.traceId, 'cafef00d');
```
Comment (kjin) [inline packages/opencensus-propagation-stackdriver/test/test-v1.ts]: nice

File: packages/opencensus-propagation-stackdriver/test/test-v1.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      input: {traceId: '123456', spanId: '667', options: undefined},
+      output: '123456/667'
+    },
+    {input: {traceId: '123456', spanId: '667'}, output: '123456/667'}
+  ];
+
+  testData.forEach(({input, output}) => {
+    it(`returns well-formatted trace context for ${inspect(input)}`, () => {
+      const setter = {
+        setHeader(name: string, value: string) {
```
Comment (kjin) [inline packages/opencensus-propagation-stackdriver/test/test-v1.ts]: Maybe assert that this function will be called?

File: packages/opencensus-propagation-stackdriver/src/v1.ts (hunk 2 above)
Comment (ofrobots) [reply packages/opencensus-propagation-stackdriver/src/v1.ts]: I don't think so. This is derived from the source file in the Stackdriver trace agent.

File: packages/opencensus-propagation-stackdriver/test/test-v1.ts (hunk 4 above)
Comment (ofrobots) [reply packages/opencensus-propagation-stackdriver/test/test-v1.ts]: Changed the tests substantially. This is no longer an issue.

File: packages/opencensus-propagation-stackdriver/test/test-v1.ts [hunk 5]
```diff
@@ -108,12 +91,47 @@ describe('inject', () => {
...
-      };
-      inject(setter, input);
+      const header = serializeSpanContext(input);
+      assert.deepEqual(header, output);
     });
   });
 });
+
+describe('extract', () => {
+  it('should return prased header from getter.getHeader', () => {
```
Comment (kjin) [inline packages/opencensus-propagation-stackdriver/test/test-v1.ts]: `parsed`

File: packages/opencensus-propagation-stackdriver/test/test-v1.ts [hunk 6]
```diff
@@ -108,12 +91,47 @@ describe('inject', () => {
...
+        done();
+      }
+    };
+    inject(setter, spanContext);
+  });
+});
+
+describe('generate', () => {
+  it('should generate a valid span context', () => {
+    const TIMES = 20;
```
Comment (kjin) [inline packages/opencensus-propagation-stackdriver/test/test-v1.ts]: Should we be checking uniqueness as well?

File: packages/opencensus-propagation-stackdriver/src/index.ts (hunk 1 above)
Comment (fabiogomessilva) [reply packages/opencensus-propagation-stackdriver/src/index.ts]: There is a SpanContext already defined in Core package. Do you any restriction to use it here?

File: packages/opencensus-propagation-stackdriver/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as v1API from './v1';
+
+export interface SpanContext {
+  /** Trace ID */
```
Comment (fabiogomessilva) [inline packages/opencensus-propagation-stackdriver/src/index.ts]: There is a SpanContext already defined in Core package. Do you have any restriction to use it here?

File: packages/opencensus-propagation-stackdriver/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * A transport and environment neutral API for setting headers.
+ */
+export interface HeaderSetter {
+  setHeader(name: string, value: string): void;
+}
+
+
+export interface Propagation {
+  extract(getter: HeaderGetter): SpanContext|null;
```
Comment (fabiogomessilva) [inline packages/opencensus-propagation-stackdriver/src/index.ts]: We were planning to add a Propagation interface to Core package in the next few weeks. What do you think about to add this interface to Core package? It should be at :
```
opencensus-core/src/trace/propagation/types.ts
```
or maybe in a new PR.

Comment (ofrobots) [reply packages/opencensus-propagation-stackdriver/src/index.ts]: Here's the list of dependencies that `core` has:

```
  "dependencies": {
    "continuation-local-storage": "^3.2.1",
    "debug": "^3.1.0",
    "intercept-stdout": "^0.1.2",
    "log-driver": "^1.2.7",
    "semver": "^5.5.0",
    "shimmer": "^1.2.0",
    "uuid": "^3.2.1"
  }
```

I think the `propagation` modules should have close to zero dependencies. In my specific use case, I want to use it in a logging context, and I don't specifically do not want to bring in `continuation-local-storage` or other similar functionality. The only things I need from `core` are two abstract types.

I think this discussion is suggesting to me that `core` is perhaps not "core-enough". /cc @kjin 

> in the next few weeks. 

I need this propagation module right now. If there is refactoring possible – I'm +1 but let's remove the code from when you're ready in a few weeks. For now this type can live here.

Comment (kjin) [reply packages/opencensus-propagation-stackdriver/src/index.ts]: > I think this discussion is suggesting to me that `core` is perhaps not "core-enough".
This makes sense to me. We should discuss the possibility of an interfaces-only module.

Comment (fabiogomessilva) [reply packages/opencensus-propagation-stackdriver/src/index.ts]: > ..  If there is refactoring possible – I'm +1 but let's remove the code from when you're ready in a few weeks. For now this type can live here.

Ok

Do you prefer to open yourself the PR to include those definitions to ``opencensus-core/src/trace/propagation/types.ts`` or do you mind if we do so? Of course, we'll gonna keep you as the author of the commit.

File: packages/opencensus-propagation-stackdriver/test/test-v1.ts (hunk 6 above)
Comment (ofrobots) [reply packages/opencensus-propagation-stackdriver/test/test-v1.ts]: Done.

**Pull request merged**